pkg/controller/baremetalhost/baremetalhost_controller_test.go

```diff
@@ -243,6 +246,224 @@ func TestSetLastUpdated(t *testing.T) {
+		func(host *metal3v1alpha1.BareMetalHost, result reconcile.Result) bool {
+			if !host.Status.Provisioning.PendingRebootSince.Before(host.Status.Provisioning.PoweredOnAt) {
+				return false
+			}
+
+			if !host.Status.PoweredOn {
+				return false
+			}
+
+			return true
+		},
+	)
+}
+
+// TestRebootWithSuffixedAnnotation tests a full reboot cycle, with suffixed annotation
+// to verify that controller holds power off until annotation removal
+func TestRebootWithSuffixedAnnotation(t *testing.T) {
+	host := newDefaultHost(t)
+	host.Annotations = make(map[string]string)
+	annotation := rebootAnnotationPrefix + "/foo"
+	host.Annotations[annotation] = ""
+	host.RecordPoweredOn()
+	secondAgo := metav1.NewTime(host.Status.Provisioning.PoweredOnAt.Add(-time.Second))
+	host.Status.Provisioning.PoweredOnAt = &secondAgo
+	host.Status.PoweredOn = true
+	host.Spec.Online = true
+
+	r := newTestReconciler(host)
+	tryReconcile(t, r, host,
+		func(host *metal3v1alpha1.BareMetalHost, result reconcile.Result) bool {
+			if host.Status.Provisioning.PendingRebootSince == nil {
+				return false
+			}
+
+			if host.Status.Provisioning.PendingRebootSince.Before(host.Status.Provisioning.PoweredOnAt) {
+				return false
+			}
+			return true
+		},
+	)
+
+	tryReconcile(t, r, host,
+		func(host *metal3v1alpha1.BareMetalHost, result reconcile.Result) bool {
+			if host.Status.Provisioning.PendingRebootSince.Before(host.Status.Provisioning.PoweredOnAt) {
+				return false
+			}
+
+			if host.Status.PoweredOn {
+				return false
+			}
+
+			return true
+		},
+	)
+
+	secondAgo = metav1.NewTime(host.Status.Provisioning.PendingRebootSince.Add(-time.Second))
+	host.Status.Provisioning.PendingRebootSince = &secondAgo
+
+	tryReconcile(t, r, host,
+		func(host *metal3v1alpha1.BareMetalHost, result reconcile.Result) bool {
+			if host.Status.Provisioning.PendingRebootSince.Before(host.Status.Provisioning.PoweredOnAt) {
+				return false
+			}
+
+			if host.Status.PoweredOn { //we expect that the machine will be powered off until we remove annotation
+				return false
+			}
+
+			return true
+		},
+	)
+
+	secondAgo = metav1.NewTime(host.Status.Provisioning.PendingRebootSince.Add(-time.Second))
+	host.Status.Provisioning.PendingRebootSince = &secondAgo
+
+	delete(host.Annotations, annotation)
+	r.client.Update(goctx.TODO(), host)
+
+	tryReconcile(t, r, host,
+		func(host *metal3v1alpha1.BareMetalHost, result reconcile.Result) bool {
+
+			if !host.Status.PoweredOn {
+				return false
+			}
+
+			if !host.Status.Provisioning.PendingRebootSince.Before(host.Status.Provisioning.PoweredOnAt) {
+				return false
+			}
+
+			return true
+		},
+	)
+
+	//make sure we don't go into another reboot
+	tryReconcile(t, r, host,
+		func(host *metal3v1alpha1.BareMetalHost, result reconcile.Result) bool {
+			if !host.Status.Provisioning.PendingRebootSince.Before(host.Status.Provisioning.PoweredOnAt) {
+				return false
+			}
+
+			if !host.Status.PoweredOn {
+				return false
+			}
+
```

Comment:
If we're going to remove the annotation, we should test for that here, too.

Reply:
In the suffixed case, the controller doesn't remove the annotation, and in the suffixless I have a check for that. let me know if you still think that something is missing.